### ISSUE TITLE
Allow generators to process extracted objects

### ISSUE THREAD

**Issue by zackw** [mesonbuild/meson#6476]
**Describe the bug**

I want to transform the object files created by a `shared_library` invocation and then put the transformed objects into a `static_library`.  I tried to do this with a generator, but meson won't let me pass the result of `extract_all_objects()` to `generator.process()`.

This appears to be related to, but not the same as, #1141 (the types involved are not the same).

**To Reproduce**

Sample meson.build that should reproduce the problem:

```
project('demo')
shlib = shared_library('demo', sources: ['demo.c'])
xfrm = generator(find_program('cp'), 
  output: '@BASENAME@.st.o',
  arguments: ['@INPUT@', '@OUTPUT@'])
statlib = static_library('demo',
  objects: xfrm.process(shlib.extract_all_objects()))
```

where `demo.c` is some trivial but nonempty C source code, e.g.

```
int define_something;
```

Running meson on the above gives this error:

```
meson.build:6:0: ERROR: Generator arguments must be strings or files not
    <Holder: <ExtractedObjects 'demo': [<File: demo.c (not built)>]>>.
```

(Of course, in my real project, the generator does more than just copy its input to its output.)

**Expected behavior**

Each of the object files for the shared library should be passed to the generator, and the resulting transformed object files should be packed into the static library.

**system parameters**

* `meson --version`: 0.53.0

**Comment by bonzini** [comment]
Error is now

```
meson.build:7:16: ERROR: generator.process argument 1 was of type "ExtractedObjects" but should have been one of: "str", "File", "CustomTarget", "CustomTargetIndex", "GeneratedList"
```

but is still present.

**Comment by bonzini** [comment]
custom_targets use determine_ext_objs() to resolve ExtractedObjects to the list of object file paths. It's not easy to do the same for generators because `process_files` currently wants to know its outputs at interpreter time (because in turn `GeneratedList.add_file()` wants to compute outputs via `@BASENAME@` and the like).

So, neither custom_target() nor generator.process() can transform each object file  individually with @BASENAME@-based output names.

**Comment by zackw** [comment]
I want to be clear that this is a hard blocker for me - without this feature, Meson is unusable for the project I wanted to use it for (libxcrypt).

**Comment by bonzini** [comment]
Thanks—can you clarify why you need to use generators and cannot use custom_target instead?

**Comment by bonzini** [comment]
(I guess that's because you want to use `@BASENAME`). If you have a partial conversion I can take a look.

**Comment by zackw** [comment]
The partial conversion can be seen at <https://github.com/besser82/libxcrypt/tree/zack/meson>. If I remember correctly, the meson.build that's there _does_ work, but fails to let me eliminate all the #defines at the bottom of crypt-port.h, which I was really hoping to replace with `objcopy` once libtool was no longer an obstacle.  The idea was to take all the `.o` files from the shared library and run them `objcopy --prefix-symbols`.  This has to be done one at a time for each object, hence the need for a generator instead of a custom target.

Also, I should clarify that the _main_ reason I gave up on the conversion is because I was told that libxcrypt having a build-time dependency on Python would cause intractable bootstrap issues for Fedora and probably also Debian.  So this isn't an _urgent_ need from my end, despite its being a blocker; but I might have pushed back on the objection if I hadn't gotten stuck on this blocker.

**Comment by bonzini** [comment]
> Also, I should clarify that the main reason I gave up on the conversion is because I was told that libxcrypt having a build-time dependency on Python would cause intractable bootstrap issues for Fedora and probably also Debian

It wouldn't be insurmountable as it would simply be part of the initial cross-compiled set of packages, but they try to keep that as small as possible indeed.

Thanks for clarifying what you want, it's a variant of `both_libraries()` that _could_ be done by Meson natively but it is also a reasonable use of generators. And I presume that, because you want the static library's objects to be usable piecewise, you can't just (for example) use `ld -r` + `objcopy`.

As I wrote earlier there are some complications because object file names are not built until much later, which is why no one ever tried looking at either this or #2657; but it's fixable.

**Comment by zackw** [comment]
> I presume that ... you want the static library's objects to be usable piecewise

Right.  Or more precisely, we want to preserve the linker's ability to discard unused objects from the static library.
